faust/tables/recovery.py

```diff
@@ -0,0 +1,236 @@
+import typing
+from collections import defaultdict
+from typing import Any, List, MutableMapping, Set, cast
+
+from mode import Service
+from mode.utils.compat import Counter
+
+from faust.types import AppT, EventT, TP
+from faust.types.tables import CollectionT, TableManagerT
+from faust.types.transports import ConsumerT
+from faust.utils import terminal
+
+if typing.TYPE_CHECKING:
+    from .manager import TableManager
+else:
+    class TableManager: ...  # noqa
+
+
+class Recovery(Service):
```

Comment:
@ask Let me see if I got this right:
* We have a single instance of `TableManager` per Faust `App`
* `app.table_manager` will have a single instance of `Recovery`, which means that the idea is for it to run for the lifetime of the application. 
* When the recovery service needs to be balanced `app.table_manager` calls `recovery.on_rebalance` which deals with active and standby topic partitions (tps), offsets and highwaters

I have a question about `_slurp_changelogs`:
* It currently is modeled as a `Service.task`, which means that it will run for as long as `Recovery` runs, right?

* Does this mean that once the table is fully recreated and up to date with the changelogs, if there is any update in the table and these are recorded to the changelog, will `event: EventT = await changelog_queue.get()` try to apply an update that has already happened?

Reply:
You got this right!  The first three points are all correct, but the next step on this patch that I'm working on now is:

`on_rebalance` executes as part of the Kafka `on_partitions_assigned` callback, so needs to return within `session.timeout` seconds (60 seconds).

We are not going to do anything in on_rebalance, we will return ASAP but first signal the recover service to:

1) In the background stop flow into agents, pause all partitions
2) Wait for all active tables to recover from changelogs
3) When that's done resume flow into agents.

This means the rebalance will not block for long, and hopefully it will eliminate rebalancing loops (errors moving across the cluster).  The simpler this process is, the less chance of a rebalancing loop.

What's missing from this patch is also reporting of progress during this process (how much is left to recover for each table).

-- 
Slurp changelogs:

Yep, this is a background service that runs ALL the time.  The `offsets` mapping will make sure we don't apply an update that has already happened.

Remember that there are not just active tables to recover changelogs for, there are also
"standbys".  If one of the nodes go down we have standby nodes ready to take over, so for standbys we constantly receive changelog events so that we can recover immediately.

Reply:
@ask Thanks for the explanation, it makes sense to optimize for simple and fast `on_rebalance`